lib/trino-hive-formats/src/main/java/io/trino/hive/formats/avro/AvroPageDataReader.java

```diff
@@ -0,0 +1,1065 @@
+import org.apache.avro.io.FastReaderBuilder;
+import org.apache.avro.io.parsing.ResolvingGrammarGenerator;
+import org.apache.avro.util.internal.Accessor;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.BiConsumer;
+import java.util.function.IntFunction;
+import java.util.stream.IntStream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Verify.verify;
+import static io.trino.hive.formats.avro.AvroTypeUtils.UNION_FIELD_TAG_TYPE;
+import static io.trino.hive.formats.avro.AvroTypeUtils.isSimpleNullableUnion;
+import static io.trino.hive.formats.avro.AvroTypeUtils.typeFromAvro;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static io.trino.spi.type.DoubleType.DOUBLE;
+import static io.trino.spi.type.IntegerType.INTEGER;
+import static io.trino.spi.type.RealType.REAL;
+import static io.trino.spi.type.VarbinaryType.VARBINARY;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.lang.Float.floatToRawIntBits;
+import static java.util.Objects.requireNonNull;
+
+public class AvroPageDataReader
+        implements DatumReader<Optional<Page>>
+{
+    private final Schema readerSchema;
+    private Schema writerSchema;
+    private final PageBuilder pageBuilder;
+    private RowBlockBuildingDecoder rowBlockBuildingDecoder;
+    private final AvroTypeManager typeManager;
+
+    public AvroPageDataReader(Schema readerSchema, AvroTypeManager typeManager)
+            throws AvroTypeException
+    {
+        try {
+            this.readerSchema = requireNonNull(readerSchema, "readerSchema is null");
+            this.typeManager = requireNonNull(typeManager, "typeManager is null");
+            this.writerSchema = this.readerSchema;
+            Type readerSchemaType = typeFromAvro(this.readerSchema, typeManager);
+            verify(readerSchemaType instanceof RowType, "Can only build pages when top level type is Row");
+            this.pageBuilder = new PageBuilder(readerSchemaType.getTypeParameters());
+            initialize();
+        }
+        catch (org.apache.avro.AvroTypeException runtimeAvroException) {
+            throw new AvroTypeException(runtimeAvroException.getMessage(), runtimeAvroException.getCause());
+        }
+    }
+
+    private void initialize()
+            throws AvroTypeException
+    {
+        verify(readerSchema.getType().equals(Schema.Type.RECORD), "Avro schema for page reader must be record");
+        verify(writerSchema.getType().equals(Schema.Type.RECORD), "File Avro schema for page reader must be record");
+        this.rowBlockBuildingDecoder = new RowBlockBuildingDecoder(writerSchema, readerSchema, typeManager);
+    }
+
+    protected static class AvroTypeExceptionRuntimeWrapper
```

Comment:
place at the bottom of this file

Reply:
Add a comment explaining that this exception only exists to propagate checked AvroTypeException through Avro interfaces that can not throw checked exceptions like `DatumReader`, and should be immediately unwrapped on the caller's side

Reply:
Rename this `UncheckedAvroTypeException` which follows the naming convention of `UncheckedIOException`